overlord/snapshotstate/snapshotmgr.go

```diff
@@ -0,0 +1,277 @@
+}
+
+func filename(setID uint64, si *snap.Info) string {
+	skel := &client.Snapshot{
+		SetID:    setID,
+		Snap:     si.Name(),
+		Revision: si.Revision,
+		Version:  si.Version,
+	}
+	return backend.Filename(skel)
+}
+
+// prepareSave does all the steps of doSave that require the state lock;
+// it has no real significance beyond making the lock handling simpler
+func prepareSave(task *state.Task) (snapshot *snapshotState, cur *snap.Info, cfg map[string]interface{}, err error) {
+	st := task.State()
+	st.Lock()
+	defer st.Unlock()
+
+	if err := task.Get("snapshot", &snapshot); err != nil {
+		return nil, nil, nil, err
+	}
+	cur, err = snapstateCurrentInfo(st, snapshot.Snap)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	snapshot.Filename = filename(snapshot.SetID, cur)
+	task.Set("snapshot", &snapshot)
+
+	rawCfg, err := configGetSnapConfig(st, snapshot.Snap)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	if rawCfg != nil {
+		if err := json.Unmarshal(*rawCfg, &cfg); err != nil {
+			return nil, nil, nil, err
+		}
+	}
+
+	return snapshot, cur, cfg, nil
+}
+
+func doSave(task *state.Task, tomb *tomb.Tomb) error {
+	snapshot, cur, cfg, err := prepareSave(task)
+	if err != nil {
+		return err
+	}
+	_, err = backendSave(tomb.Context(nil), snapshot.SetID, cur, cfg, snapshot.Users)
+	return err
+}
+
+// prepareRestore does the steps of doRestore that require the state lock
+// before the backend Restore call.
+func prepareRestore(task *state.Task) (snapshot *snapshotState, oldCfg map[string]interface{}, reader *backend.Reader, err error) {
+	st := task.State()
+
+	st.Lock()
+	defer st.Unlock()
+
+	if err := task.Get("snapshot", &snapshot); err != nil {
+		return nil, nil, nil, err
+	}
+
+	rawCfg, err := configGetSnapConfig(st, snapshot.Snap)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	if rawCfg != nil {
+		if err := json.Unmarshal(*rawCfg, &oldCfg); err != nil {
+			return nil, nil, nil, err
+		}
+	}
+
+	reader, err = backendOpen(snapshot.Filename)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	return snapshot, oldCfg, reader, nil
+}
+
+func doRestore(task *state.Task, tomb *tomb.Tomb) error {
+	snapshot, oldCfg, reader, err := prepareRestore(task)
+	if err != nil {
+		return err
+	}
+
+	restoreState, err := backendRestore(reader, tomb.Context(nil), snapshot.Users, task.Logf)
+	if err != nil {
+		return err
+	}
+
+	buf, err := json.Marshal(reader.Conf)
+	if err != nil {
+		backendRevert(restoreState)
+		return err
+	}
+
+	st := task.State()
+	st.Lock()
+	defer st.Unlock()
+
+	if err := configSetSnapConfig(st, snapshot.Snap, (*json.RawMessage)(&buf)); err != nil {
+		backendRevert(restoreState)
+		return err
+	}
+
+	restoreState.Config = oldCfg
+	task.Set("restore-state", restoreState)
+
+	return nil
+}
+
+func undoRestore(t *state.Task, _ *tomb.Tomb) error {
+	var restoreState backend.RestoreState
+	var snapshot snapshotState
+
+	st := t.State()
+	st.Lock()
+	defer st.Unlock()
+
+	t.Get("restore-state", &restoreState)
```

Comment:
Even though we know we set these values on the task in doRestore, we should probably check for errors as we do in all existing undo handlers.